### ISSUE TITLE
Cursor disappears when viewing error message

### ISSUE THREAD

**Issue by Skyb0rg007** [w0rp/ale#1334]
## Information

**VIM 8.0: vimx package**

VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Feb  1 2018 14:51:52)
Included patches: 1-1451

Operating System: Fedora 27

**NOTE:** This issue only exists on the terminal version of vimx (vim-X11)
ie. This bug does not exist on the version of vim pre-installed in Fedora 27
nor does it exist in gVim (which comes in the same package)

### :ALEInfo

 Current Filetype: c
Available Linters: ['clang', 'clangtidy', 'cppcheck', 'gcc']
  Enabled Linters: ['clang', 'clangtidy', 'cppcheck', 'gcc']
 Linter Variables:

let g:ale_c_clang_executable = 'clang'
let g:ale_c_clang_options = '-std=c11 -Wall'
let g:ale_c_clangtidy_checks = ['*']
let g:ale_c_clangtidy_executable = 'clang-tidy'
let g:ale_c_clangtidy_options = ''
let g:ale_c_cppcheck_executable = 'cppcheck'
let g:ale_c_cppcheck_options = '--enable=style'
let g:ale_c_gcc_executable = 'gcc'
let g:ale_c_gcc_options = '-std=c11 -Wall'
 Global Variables:

let g:ale_cache_executable_check_failures = 0
let g:ale_change_sign_column_color = 0
let g:ale_command_wrapper = ''
let g:ale_completion_delay = 100
let g:ale_completion_enabled = 0
let g:ale_completion_max_suggestions = 50
let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'E'
let g:ale_echo_msg_format = '[%linter%] %s'
let g:ale_echo_msg_info_str = 'Info'
let g:ale_echo_msg_warning_str = 'W'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {}
let g:ale_history_enabled = 1
let g:ale_history_log_output = 1
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'always'
let g:ale_lint_on_insert_leave = 0
let g:ale_linter_aliases = {}
let g:ale_linters = {}
let g:ale_linters_explicit = 0
let g:ale_list_window_size = 10
let g:ale_loclist_msg_format = '[%linter%] %s'
let g:ale_max_buffer_history_size = 20
let g:ale_max_signs = -1
let g:ale_maximum_file_size = 0
let g:ale_open_list = 0
let g:ale_pattern_options = {}
let g:ale_pattern_options_enabled = 0
let g:ale_set_balloons = 1
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_info = '--'
let g:ale_sign_offset = 1000000
let g:ale_sign_style_error = '>>'
let g:ale_sign_style_warning = '--'
let g:ale_sign_warning = '--'
let g:ale_statusline_format = ['%d error(s)', '%d warning(s)', 'OK']
let g:ale_type_map = {}
let g:ale_warn_about_trailing_blank_lines = 1
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

(executable check - success) clang
(started) ['/bin/bash', '-c', '''clang'' -S -x c -fsyntax-only -iquote ''/home/ssoss/Documents/Programming/C'' -std=c11 -Wall - < ''/tmp/vsa7l4x/2/Dynamic.c''']
(executable check - failure) clang-tidy
(executable check - failure) cppcheck
(executable check - success) gcc
(started) ['/bin/bash', '-c', '''gcc'' -S -x c -fsyntax-only -iquote ''/home/ssoss/Documents/Programming/C'' -std=c11 -Wall - < ''/tmp/vsa7l4x/3/Dynamic.c''']
(finished - exit code 0) ['/bin/bash', '-c', '''clang'' -S -x c -fsyntax-only -iquote ''/home/ssoss/Documents/Programming/C'' -std=c11 -Wall - < ''/tmp/vsa7l4x/4/Dynamic.c''']

<<<NO OUTPUT RETURNED>>>

(executable check - failure) clang-tidy
(executable check - failure) cppcheck
(finished - exit code 0) ['/bin/bash', '-c', '''gcc'' -S -x c -fsyntax-only -iquote ''/home/ssoss/Documents/Programming/C'' -std=c11 -Wall - < ''/tmp/vsa7l4x/5/Dynamic.c''']

<<<NO OUTPUT RETURNED>>>

(finished - exit code 1) ['/bin/bash', '-c', '''clang'' -S -x c -fsyntax-only -iquote ''/home/ssoss/Documents/Programming/C'' -std=c11 -Wall - < ''/tmp/vsa7l4x/9/Dynamic.c''']

<<<OUTPUT STARTS>>>
<stdin>:11:1: error: unknown type name 'iansd'
iansd
^
<stdin>:12:1: error: expected identifier or '('
describe({
^
<stdin>:8:5: note: expanded from macro 'describe'
    void concat(test_, __COUNTER__)() { \
    ^
2 errors generated.
<<<OUTPUT ENDS>>>

(executable check - failure) cppcheck
(finished - exit code 1) ['/bin/bash', '-c', '''gcc'' -S -x c -fsyntax-only -iquote ''/home/ssoss/Documents/Programming/C'' -std=c11 -Wall - < ''/tmp/vsa7l4x/10/Dynamic.c''']

<<<OUTPUT STARTS>>>
<stdin>:8:5: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ before ‘void’
<stdin>:12:1: note: in expansion of macro ‘describe’
<<<OUTPUT ENDS>>>

(finished - exit code 0) ['/bin/bash', '-c', '''clang'' -S -x c -fsyntax-only -iquote ''/home/ssoss/Documents/Programming/C'' -std=c11 -Wall - < ''/tmp/vsa7l4x/12/Dynamic.c''']

<<<NO OUTPUT RETURNED>>>

(executable check - failure) cppcheck
(finished - exit code 0) ['/bin/bash', '-c', '''gcc'' -S -x c -fsyntax-only -iquote ''/home/ssoss/Documents/Programming/C'' -std=c11 -Wall - < ''/tmp/vsa7l4x/13/Dynamic.c''']

<<<NO OUTPUT RETURNED>>>

## What went wrong

When there is an ALE error, the cursor disappears in normal mode for as long as there is an error shown in the command bar

## Reproducing the bug

Steps for repeating the bug:

1. Install the vim-X11 package on Fedora 27
2. Using GNOME terminal or Terminator, launch vimx and edit a file that has linting enabled\*
3. Enter normal mode after an error is shown, and move the cursor to that line
4. Cursor disappears until the error times out

\* The bug definitely exists on Terminator and GNOME terminal
\*\* There is no need to configure your .vimrc other than install ALE

Other links:
**Imgur of the bug:** https://imgur.com/a/QYy5R

**Comment by stefan-squareweave** [comment]
Can confirm, also running Fedora. The problem seems to be related to the "EchoImpl" function called when you move the cursor over a line with an error (or off it). If I comment out autoload/ale/cursor.vim:22 and autoload/ale/cursor.vim:82 on revision a3329ef3fc44a04a2617941eb46deef6c184f4af then the problem goes away (and so does the functionality). Not sure of how to correct those lines though.

A workaround to stop the issue is to suppress the display of errors when you move the cursor onto their line using: `let g:ale_echo_cursor = 0` .

**Comment by tuomastiainen** [comment]
I am having the same issue, running Vim version 8.0.1473 on Ubuntu 16.04, on with Tmux on Gnome Terminal emulator. I also used the workaround to disable to echoing of errors.

**Comment by NotAFile** [comment]
Is this the fault of the `vimx` package or the `ale` package? Has this been reported to the fedora vimx maintainers?

**Comment by stefan-squareweave** [comment]
After your comment NotAFile, I tested out the vim-enhanced package on Fedora 27 (normally I use vimx), and there's no issue there. tuomastianinen experienced it on the same version running under Ubuntu (unspecified whether it's vimx or not).

These data suggest that the most likely culprit is an interaction with the X11 compilation options for vim. Whether it's ALE using an API in a strange way, or a breakage caused by X11 integration is not certain yet.

I guess the next thing to do would be to extract the code out of ALE that triggers the issue, confirm it happens on a couple of distros, and if so, submit a bug to upstream vim.

**Comment by jcsjcs** [comment]
I had the same problem on Ubunu 16.04. The latest update to vim has corrected the behaviour.
This version of vim works for me:
```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Feb  8 2018 10:52:04)                                                                            
Included patches: 1-1476

```

**Comment by inducer** [comment]
8.0.0197 also does not seem to exhibit the issue.

**Comment by NotAFile** [comment]
Occurs for me on:

```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Feb  5 2018 12:21:51)
Included patches: 1-1473
```

I guess one of those few patches are what fixed it?

**Comment by ollien** [comment]
`vimx-2:8.0.1176-1.fc27` does not exhibit the issue, but 1473 does.

**Comment by tuomastiainen** [comment]
The issue was also fixed for me when updating Vim from 1-1473 to 1-1481.

**Comment by Skyb0rg007** [comment]
The issue was fixed for me on 1478

**Comment by w0rp** [comment]
Sounds like this was a temporary Vim bug. If anyone else has this problem, try a different Vim version, and that will probably fix it.

**Comment by arkgast** [comment]
The error persists in this version.

 VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Feb 24 2018 10:05:30)
Included patches: 1-1534

**Comment by kmARC** [comment]
I'd recommend implementing a workaround for this bug, since the latest LTS version of Ubuntu (18.04) comes with a vim version that is affected.

Please consider resolving this issue if you can, thanks

```
[~] apt-cache show vim-common
Package: vim-common
Architecture: all
Version: 2:8.0.1453-1ubuntu1
Priority: important
Section: editors
Source: vim
Origin: Ubuntu
Maintainer: Ubuntu Developers <ubuntu-devel-discuss@lists.ubuntu.com>
Original-Maintainer: Debian Vim Maintainers <pkg-vim-maintainers@lists.alioth.debian.org>
Bugs: https://bugs.launchpad.net/ubuntu/+filebug
Installed-Size: 329
Depends: xxd
Recommends: vim | vim-gtk | vim-gtk3 | vim-athena | vim-nox | vim-tiny
Filename: pool/main/v/vim/vim-common_8.0.1453-1ubuntu1_all.deb
Size: 70396
MD5sum: c2df8446f4a6545cd1cf7406d9dcc0a2
SHA1: 204a6deab866c64e58e74960dce6e14d3dde0748
SHA256: 3a1e31b1fc976cb4e6410f1860db3226ceddc373c7513443e208da256c8bf376
Homepage: https://vim.sourceforge.io/
Description-en: Vi IMproved - Common files
 Vim is an almost compatible version of the UNIX editor Vi.
 .
 This package contains files shared by all non GUI-enabled vim variants
 available in Debian.  Examples of such shared files are: manpages and
 configuration files.
Description-md5: 30b64c3dceccc4faef229eafc38d0069
Task: minimal
Supported: 5y
```

**Comment by w0rp** [comment]
Yes, the latest version of LTS version of Ubuntu comes with a version of Vim that has this bug in it. The Ubuntu package maintainers need to update the Vim package to fix it.

**Comment by mttbx** [comment]
I also suggest author implement a workaround so that people don't need to change their version.

**Comment by w0rp** [comment]
It's not possible to implement a workaround. The only solution is for the Ubuntu package managers to update the package for Vim so it doesn't include the bug.

**Comment by NotAFile** [comment]
Would setting `g:ale_echo_cursor = 0` if the user is on the broken version not be a workaround? The question is really which of the two is better, though.

**Comment by w0rp** [comment]
You can help by tagging the bug in Launchpad, which might help a package maintainer notice it and update the package. https://bugs.launchpad.net/ubuntu/+source/vim/+bug/1768026 The bug isn't in slightly earlier or later versions of Vim.

**Comment by jnreynoso** [comment]
+1

**Comment by fabriziobertoglio1987** [comment]
> I confirm this happens too on vim shipped with ubuntu 18.04 (Patch version 1453). I had to uninstall vim and use the version from this PPA : https://launchpad.net/~jonathonf/+archive/ubuntu/vim

I am also running Ubuntu 18.04 with Vim 8.0.1453  

Installing the pachage from @w0rp solved the problem
Alternative is [following the guide at vim official page](https://www.vim.org/git.php) to install vim from source. You can clone the [latest release from github](https://github.com/vim/vim/releases)

**Comment by Henry-E** [comment]
Also running into this bug, was hoping to try ALE but it's back to syntastic I guess

**Comment by fabriziobertoglio1987** [comment]
@Henry-E 

I was able to solve this. 
I don't remember well, but I believe it was related to the vim interface colors plugins 

This are my current settings

```
" Interface
Plugin 'morhetz/gruvbox'
" Plugin 'ayu-theme/ayu-vim'
" Plugin 'drewtempelmeyer/palenight.vim'

" gruvbox
set background=dark
" set background=light
let g:gruvbox_contrast_light = 'hard'

" ayu-vim
" set termguicolors     " enable true colors support
" let ayucolor="light"  " for light version of theme
" let ayucolor="mirage" " for mirage version of theme
" let ayucolor="dark"   " for dark version of theme
" colorscheme ayu
```

![Screenshot from 2019-03-22 11-53-03](https://user-images.githubusercontent.com/24992535/54818344-5322af00-4c99-11e9-806f-1a78addf72f0.png)

**Comment by Henry-E** [comment]
Great thanks! Installing gvim is definitely easier than doing installing from source as I also need clipboard support. I'll give it a shot

**Comment by fabriziobertoglio1987** [comment]
@Henry-E 
sorry .. I just re-enabled the plugin and I also currently have the same problem with gvim as it is the same version Vim 8.0.1453, which is shipping the bug 

> I confirm this happens too on vim shipped with ubuntu 18.04 (Patch version 1453). I had to uninstall vim and use the version from this PPA : https://launchpad.net/~jonathonf/+archive/ubuntu/vim

that ppa could be solving the issue, I remember experiencing some other bugs .. and currently I am using 

```
let g:ale_echo_cursor = 0
```
in my vimrc which seems to have solved the issue

**Comment by banderlog** [comment]
In Ubuntu 18.04 vim is still unpatched, thus I installed one from Jonathon F. PPA:

```
sudo apt-get install software-properties-common
sudo add-apt-repository ppa:jonathonf/vim
sudo apt install vim
```

All works now

**Comment by w0rp** [comment]
I use that PPA myself. That's my personal recommendation. :+1:

**Comment by fabriziobertoglio1987** [comment]
@w0rp the ppa does not have integration with the clipboard.
You can check by running `vim -V`, it will output on ubuntu  `- clipboard`, which means no clipboard integration.

Ubuntu users need to install `vim-gnome` to have the clipboard integration

```
sudo apt install vim-gnome
```

Ubuntu `18.04` users will install version `8.0.1453` with vim-gnome package as [showcased here](https://packages.ubuntu.com/bionic/vim-gnome), while Ubuntu 19.04 provides the updated [version `8.1.0320`](https://packages.ubuntu.com/disco/vim-gnome)

I and all the other users in this repository experienced this bug on vim `8.01453`. I compiled vim from source in its latest release and did not experience the bug (but had not clipboard integration). 

I believe that the issue is solved on `8.1.0320` and I would suggest to upgrade to ubuntu `19.04`

**Comment by KaMoKuMo** [comment]
As I didn't feel like changing my vim i put the following lines into my vimrc:
`function! SilenceWariningAle()
  let g:ale_echo_cursor=0
endfunction
function! EnableWarningAle()
  let g:ale_echo_cursor=1
endfunction

nnoremap \sw :call SilenceWarningAle()<CR>
nnoremap \ew :call EnableWarningAle()<CR>`

So..whenever my cursor vanishes i read the message and hit
`\sw`
and fix the error. Next time i need to read an error I hit
`\ew`
I know it's not pretty, but at least I'm back to work, until I can upgrade my ubuntu...
